### DISCUSSION TITLE
Jetpack Compose's CamelCase guideline for top-level consts

### DISCUSSION THREAD

**Discussion by drinkthestars** [detekt/detekt#4078] [Q&A]
Given that [Jetpack Compose is now past stable `1.0`](https://android-developers.googleblog.com/2021/07/jetpack-compose-announcement.html) and prescribes `CamelCase` for top-level consts ([here](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-api-guidelines.md#singletons-constants-sealed-class-and-enum-class-values)), what's the recommended way of dealing with this given existing `TopLevelPropertyNaming -> constantPattern` rule with default  `[A-Z][_A-Z0-9]*` ?

i.e. guideline suggests:

``` Kotlin
// Do
const val DefaultKeyName = "__defaultKey"

// Don't
const val DEFAULT_KEY_NAME = "__defaultKey"
```

**Comment by cortinico** [comment]
I'm leaning towards updating `constantPattern` to be `[A-Z][_A-Za-z0-9]*` given the wider usages of top level constant in Compose. I'd like to get others' opinion on this matter, specifically given also #3894

**Comment by schalkms** [comment]
I made by point. I think we should continue to try and stick with the official style guides.

> I also prefer to keep it, since this config change would be kind of controversial as of now. If the official style guide is updated, we can still adapt the config later on.

https://github.com/detekt/detekt/issues/3147#issuecomment-712446581

**Comment by drinkthestars** [comment]
Definitely understand the idea behind keeping base detekt adhere to Kotlin lang style only, but then having framework/platform specific stuff be a separate rule set/offering. Sounds like a larger decision to make. 

Until we can arrive at a decision + implementation there, are we leaving it up to each consumer to decide if they want to customize the rule? 

Related: resurfacing comment https://github.com/detekt/detekt/issues/3147#issuecomment-797377600

> Should Detekt provides docs about what rules are necessary to change to avoid problems with the Compose guidelines?

> I understand the default behavior of Detekt should be following the Kotlinlang guidelines, but having a doc which explain how to use specific rules (and what rules) for Compose should be great.

**Reply by cortinico** [reply]
> but having a doc which explain how to use specific rules (and what rules) for Compose should be great.

Agree. We should have a page for that. Do you want to kick it off with a PR? 🙏  We can follow up on that

**Reply by drinkthestars** [reply]
Sounds good, will do 🙏🏼

**Comment by BraisGabin** [comment]
This is kind of related with #3150. In that PR we decided to follow IntelliJ implementation. But now we are talking about following the code style. So, if we want to be consistent we should follow IntelliJ and they allow CamelCase in top-level consts. But I'd vote to stick with the kotlin style guide. Then we can create documentation for how to change our configuration to follow the Composable style.

**Reply by drinkthestars** [reply]
Yeah, I'll look into drafting a PR for the doc part of this

**Comment by drinkthestars** [comment]
Just to connect the dots further, the `MagicNumber` rule also comes into play, most commonly [in a case where we define `Color`s](https://github.com/android/compose-samples/blob/2488fe1c1df6f7f516b0030f0ed8272082151361/Owl/app/src/main/java/com/example/owl/ui/theme/Color.kt#L24-L27) in Compose:

``` Kotlin
val yellow200 = Color(0xffffeb46)
val yellow400 = Color(0xffffc000)
val yellow500 = Color(0xffffde03)
val yellowDarkPrimary = Color(0xff242316)
```

See [Add regexp to MagicNumber rule #4080](https://github.com/detekt/detekt/issues/4080) for more details. As per the [comments](https://github.com/detekt/detekt/issues/4080#issuecomment-909421128) there, though, it could relate to [Implement a more advanced way to ignore/suppress issues #4068](
https://github.com/detekt/detekt/issues/4068).

Looking into including such details into Compose-specific docs so that info is in one place.

**Reply by cortinico** [reply]
That was partially discussed also here https://github.com/detekt/detekt/issues/4080
